File: Samples/360VideoPlayback/cpp/Content/VideoRenderer.cpp
#include "pch.h"
#include "VideoRenderer.h"
#include "Common\DirectXHelper.h"
#include "Windows.Graphics.DirectX.Direct3D11.interop.h"
#include "AppView.h"

using namespace _360VideoPlayback;
using namespace concurrency;
using namespace DirectX;
using namespace Microsoft::WRL;
using namespace std::placeholders;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Numerics;
using namespace Windows::Graphics::DirectX::Direct3D11;
using namespace Windows::Media::Core;
using namespace Windows::Media::Playback;
using namespace Windows::UI::Input::Spatial;

// Loads vertex and pixel shaders from files and instantiates the cube geometry.
VideoRenderer::VideoRenderer()

{
    m_deviceResources = _360VideoPlaybackMain::GetDeviceResources();
    const CD3D11_BUFFER_DESC constantBufferDesc(sizeof(ModelConstantBuffer), D3D11_BIND_CONSTANT_BUFFER);
    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateBuffer(
            &constantBufferDesc,
            nullptr,
            &m_modelConstantBuffer));
}

// Called once per frame.
void VideoRenderer::Update(const DX::StepTimer& timer)
{
    float3 scale = 10.0f;
    const XMMATRIX modelScale = XMMatrixScalingFromVector(XMLoadFloat3(&scale));
    const XMMATRIX localTransform = modelScale;
    XMStoreFloat4x4(&m_modelConstantBufferData.model, XMMatrixTranspose(localTransform));
}

// Renders one frame using the vertex and pixel shaders.
// On devices that do not support the D3D11_FEATURE_D3D11_OPTIONS3::
// VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature,
// a pass-through geometry shader is also used to set the render
// target array index.
void VideoRenderer::Render()
{
    critical_section::scoped_lock lock(m_critical);
    // Loading is asynchronous. Resources must be created before drawing can occur.
    if (!m_loadingComplete)
    {
        return;
    }

    // Use the D3D device context to update Direct3D device-based resources.
    const auto context = m_deviceResources->GetD3DDeviceContext();

    // Update the model transform buffer for the hologram.
    context->UpdateSubresource(
        m_modelConstantBuffer.Get(),
        0,
        nullptr,
        &m_modelConstantBufferData,
        0,
        0
    );


    // Each vertex is one instance of the VertexPositionColor struct.
    const UINT stride = sizeof(VertexPositionTexture);
    const UINT offset = 0;
    context->IASetVertexBuffers(
        0,
        1,
        m_vertexBuffer.GetAddressOf(),
        &stride,
        &offset
    );
    context->IASetIndexBuffer(
        m_indexBuffer.Get(),
        DXGI_FORMAT_R16_UINT, // Each index is one 16-bit unsigned integer (short).
        0
    );
    context->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
    context->IASetInputLayout(m_inputLayout.Get());

    // Attach the vertex shader.
    context->VSSetShader(
        m_vertexShader.Get(),
        nullptr,
        0
    );
    // Apply the model constant buffer to the vertex shader.
    context->VSSetConstantBuffers(
        0,
        1,
        m_modelConstantBuffer.GetAddressOf()
    );

    if (!m_usingVprtShaders)
    {
        // On devices that do not support the D3D11_FEATURE_D3D11_OPTIONS3::
        // VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature,
        // a pass-through geometry shader is used to set the render target
        // array index.
        context->GSSetShader(
            m_geometryShader.Get(),
            nullptr,
            0
        );
    }

    // Attach the pixel shader.
    context->PSSetShader(
        m_pixelShader.Get(),
        nullptr,
        0
    );

    // Set the Texture Shader resource and samplers
    context->PSSetShaderResources(0, 1, m_textureView.GetAddressOf());
    context->PSSetSamplers(
        0,
        1,
        m_quadTextureSamplerState.GetAddressOf()
    );


    // Draw the objects.
    context->DrawIndexedInstanced(
        m_indexCount,   // Index count per instance.
        2,              // Instance count.
        0,              // Start index location.
        0,              // Base vertex location.
        0               // Start instance location.
    );

}

void VideoRenderer::CreateDeviceDependentResources()
{
    m_usingVprtShaders = m_deviceResources->GetDeviceSupportsVprt();

    // Create the Texture, ShaderResource and Sampler state

    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateTexture2D(
            &CD3D11_TEXTURE2D_DESC(
                DXGI_FORMAT_R8G8B8A8_UNORM,
                AppView::GetMediaPlayer()->PlaybackSession->NaturalVideoWidth,        // Width
                AppView::GetMediaPlayer()->PlaybackSession->NaturalVideoHeight,        // Height
                1,          // MipLevels
                1,          // ArraySize
                D3D11_BIND_SHADER_RESOURCE | D3D11_BIND_RENDER_TARGET
            ),
            nullptr,
            &m_texture
        )
    );

    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateShaderResourceView(
            m_texture.Get(), nullptr,
            &m_textureView
        )
    );


    D3D11_SAMPLER_DESC desc;
    ZeroMemory(&desc, sizeof(D3D11_SAMPLER_DESC));


    desc.Filter = D3D11_FILTER_ANISOTROPIC;
    desc.AddressU = D3D11_TEXTURE_ADDRESS_CLAMP;
    desc.AddressV = D3D11_TEXTURE_ADDRESS_CLAMP;
    desc.AddressW = D3D11_TEXTURE_ADDRESS_CLAMP;
    desc.MaxAnisotropy = 3;
    desc.MinLOD = 0;
    desc.MaxLOD = 3;
    desc.MipLODBias = 0.f;
    desc.BorderColor[0] = 0.f;
    desc.BorderColor[1] = 0.f;
    desc.BorderColor[2] = 0.f;
    desc.BorderColor[3] = 0.f;
    desc.ComparisonFunc = D3D11_COMPARISON_NEVER;

    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateSamplerState(
            &desc,
            &m_quadTextureSamplerState
        )
    );

    CreateD3D11Surface();
    LoadShaders();
}

void VideoRenderer::LoadShaders()
{
    // On devices that do support the D3D11_FEATURE_D3D11_OPTIONS3::
    // VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature
    // we can avoid using a pass-through geometry shader to set the render
    // target array index, thus avoiding any overhead that would be
    // incurred by setting the geometry shader stage.
    std::wstring vertexShaderFileName = m_usingVprtShaders ? L"ms-appx:///VprtVertexShader.cso" : L"ms-appx:///VertexShader.cso";

    // Load shaders asynchronously.
    task<std::vector<byte>> loadVSTask = DX::ReadDataAsync(vertexShaderFileName);
    task<std::vector<byte>> loadPSTask = DX::ReadDataAsync(L"ms-appx:///PixelShader.cso");

    task<std::vector<byte>> loadGSTask;
    if (!m_usingVprtShaders)
    {
        // Load the pass-through geometry shader.
        loadGSTask = DX::ReadDataAsync(L"ms-appx:///GeometryShader.cso");
    }

    // After the vertex shader file is loaded, create the shader and input layout.
    task<void> createVSTask = loadVSTask.then([this](const std::vector<byte>& fileData)
    {
        DX::ThrowIfFailed(
            m_deviceResources->GetD3DDevice()->CreateVertexShader(
                fileData.data(),
                fileData.size(),
                nullptr,
                &m_vertexShader
            )
        );

        constexpr std::array<D3D11_INPUT_ELEMENT_DESC, 2> vertexDesc =
        { {
            { "POSITION", 0, DXGI_FORMAT_R32G32B32_FLOAT, 0,  0, D3D11_INPUT_PER_VERTEX_DATA, 0 },
                        { "TEXCOORD", 1, DXGI_FORMAT_R32G32_FLOAT,    0, 12, D3D11_INPUT_PER_VERTEX_DATA, 0 },
        } };

        DX::ThrowIfFailed(
            m_deviceResources->GetD3DDevice()->CreateInputLayout(
                vertexDesc.data(),
                static_cast<UINT>(vertexDesc.size()),
                fileData.data(),
                static_cast<UINT>(fileData.size()),
                &m_inputLayout
            )
        );
    });

    // After the pixel shader file is loaded, create the shader and constant buffer.
    task<void> createPSTask = loadPSTask.then([this](const std::vector<byte>& fileData)
    {
        DX::ThrowIfFailed(
            m_deviceResources->GetD3DDevice()->CreatePixelShader(
                fileData.data(),
                fileData.size(),
                nullptr,
                &m_pixelShader
            )
        );


    });

    task<void> createGSTask;
    if (!m_usingVprtShaders)
    {
        // After the pass-through geometry shader file is loaded, create the shader.
        createGSTask = loadGSTask.then([this](const std::vector<byte>& fileData)
        {
            DX::ThrowIfFailed(
                m_deviceResources->GetD3DDevice()->CreateGeometryShader(
                    fileData.data(),
                    fileData.size(),
                    nullptr,
                    &m_geometryShader
                )
            );
        });
    }

    // Once all shaders are loaded, create the mesh.
    task<void> shaderTaskGroup = m_usingVprtShaders ? (createPSTask && createVSTask) : (createPSTask && createVSTask && createGSTask);
    task<void> createSphereTask = shaderTaskGroup.then([this]()
    {
        ComputeSphere(16, true);
    });

    // Once the Sphere is created, the object is ready to be rendered.
    createSphereTask.then([this]()
    {
        m_loadingComplete = true;
    });
}

void VideoRenderer::ReleaseDeviceDependentResources()
{
    m_loadingComplete = false;
    m_usingVprtShaders = false;
    m_vertexShader.Reset();
    m_inputLayout.Reset();
    m_pixelShader.Reset();
    m_geometryShader.Reset();
    m_modelConstantBuffer.Reset();
    m_vertexBuffer.Reset();
    m_indexBuffer.Reset();
    if (m_videoFrameAvailToken.Value)
    {
        if (AppView::GetMediaPlayer() != nullptr)
        {
            AppView::GetMediaPlayer()->VideoFrameAvailable -= m_videoFrameAvailToken;
        }
    }
}

void VideoRenderer::ComputeSphere(unsigned short tessellation, bool invertn)
{
    if (tessellation < 3)
        throw std::out_of_range("tesselation parameter out of range");

    unsigned short verticalSegments = tessellation;
    unsigned short horizontalSegments = tessellation * 2;
    const unsigned short vertCount = (verticalSegments + 1) * (horizontalSegments + 1);
    unsigned short indexCount = (verticalSegments) * (horizontalSegments + 1) * 6;

    float radius = 0.5f; // Diameter of the default Sphere will always be 1 to stay aligned

    std::vector<VertexPositionTexture> sphereVertices;

    for (unsigned short i = 0; i <= verticalSegments; i++)
    {
        float v = 1 - (float)i / verticalSegments;

        float latitude = (i * XM_PI / verticalSegments) - XM_PIDIV2;
        float dy, dxz;

        XMScalarSinCos(&dy, &dxz, latitude);

        // Create a single ring of vertices at this latitude.
        for (unsigned short j = 0; j <= horizontalSegments; j++)
        {
            float u = (float)j / horizontalSegments;

            float longitude = j * XM_2PI / horizontalSegments;
            float dx, dz;

            XMScalarSinCos(&dx, &dz, longitude);

            dx *= dxz;
            dz *= dxz;

            XMVECTOR normal = XMVectorSet(dx, dy, dz, 0) * radius;
            XMFLOAT3 normalFloat3;
            XMStoreFloat3(&normalFloat3, normal);
            VertexPositionTexture vert;
            vert.position = normalFloat3;

            vert.textureCoordinate = XMFLOAT2(1.0f - u, v);

            sphereVertices.push_back(vert);
        }
    }


    std::vector<unsigned short> sphereIndices;

    // Fill the index buffer with triangles joining each pair of latitude rings.
    unsigned short stride = horizontalSegments + 1;
    for (unsigned short i = 0; i < verticalSegments; i++)
    {
        for (unsigned short j = 0; j <= horizontalSegments; j++)
        {
            unsigned short nextI = i + 1;
            unsigned short nextJ = (j + 1) % stride;
            if (!invertn)
            {
                sphereIndices.push_back(i * stride + j);
                sphereIndices.push_back(nextI * stride + j);
                sphereIndices.push_back(i * stride + nextJ);

                sphereIndices.push_back(i * stride + nextJ);
                sphereIndices.push_back(nextI * stride + j);
                sphereIndices.push_back(nextI * stride + nextJ);
            }
            else
            {
                sphereIndices.push_back(nextI * stride + j);
                sphereIndices.push_back(i * stride + j);
                sphereIndices.push_back(i * stride + nextJ);

                sphereIndices.push_back(nextI * stride + j);
                sphereIndices.push_back(i * stride + nextJ);
                sphereIndices.push_back(nextI * stride + nextJ);
            }
        }
    }

    D3D11_SUBRESOURCE_DATA vertexBufferData = { 0 };
    vertexBufferData.pSysMem = sphereVertices.data();
    vertexBufferData.SysMemPitch = 0;
    vertexBufferData.SysMemSlicePitch = 0;
    const CD3D11_BUFFER_DESC vertexBufferDesc(sizeof(VertexPositionTexture) * vertCount, D3D11_BIND_VERTEX_BUFFER);
    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateBuffer(
            &vertexBufferDesc,
            &vertexBufferData,
            &m_vertexBuffer
        )
    );

    m_indexCount = indexCount;

    D3D11_SUBRESOURCE_DATA indexBufferData = { 0 };
    indexBufferData.pSysMem = sphereIndices.data();
    indexBufferData.SysMemPitch = 0;
    indexBufferData.SysMemSlicePitch = 0;
    CD3D11_BUFFER_DESC indexBufferDesc(sizeof(unsigned short) * indexCount, D3D11_BIND_INDEX_BUFFER);
    DX::ThrowIfFailed(
        m_deviceResources->GetD3DDevice()->CreateBuffer(
            &indexBufferDesc,
            &indexBufferData,
            &m_indexBuffer
        )
    );
}

void VideoRenderer::CreateD3D11Surface()
{
    Microsoft::WRL::ComPtr<IDXGISurface> spDXGIInterfaceAccess;
    DX::ThrowIfFailed(m_texture->QueryInterface(IID_PPV_ARGS(&spDXGIInterfaceAccess)));
    m_surface = CreateDirect3DSurface(spDXGIInterfaceAccess.Get());
    m_videoFrameAvailToken =
        AppView::GetMediaPlayer()->VideoFrameAvailable +=
        ref new TypedEventHandler<MediaPlayer^, Platform::Object^>(
            std::bind(&VideoRenderer::OnVideoFrameAvailable, this));
    AppView::GetMediaPlayer()->Play();
}

void VideoRenderer::OnVideoFrameAvailable()
{
    critical_section::scoped_lock lock(m_critical);
    AppView::GetMediaPlayer()->CopyFrameToVideoSurface(m_surface);
}

File: Samples/360VideoPlayback/cs/Content/VideoRenderer.cs
using _360VideoPlayback.Common;
using System;
using System.Numerics;
using System.Runtime.InteropServices;
using Windows.Graphics.DirectX.Direct3D11;
using Windows.Media.Playback;

namespace _360VideoPlayback.Content
{
    /// <summary>
    /// This sample renderer instantiates a basic rendering pipeline.
    /// </summary>
    internal class VideoRenderer : Disposer
    {
        // Cached reference to device resources.
        private DeviceResources deviceResources;

        // Direct3D resources for cube geometry.
        private SharpDX.Direct3D11.InputLayout inputLayout;
        private SharpDX.Direct3D11.Buffer vertexBuffer;
        private SharpDX.Direct3D11.Buffer indexBuffer;
        private SharpDX.Direct3D11.VertexShader vertexShader;
        private SharpDX.Direct3D11.GeometryShader geometryShader;
        private SharpDX.Direct3D11.PixelShader pixelShader;
        private SharpDX.Direct3D11.Buffer modelConstantBuffer;

        private SharpDX.Direct3D11.Texture2D m_texture;
        private SharpDX.Direct3D11.ShaderResourceView m_textureView;
        private SharpDX.Direct3D11.SamplerState m_quadTextureSamplerState;
        private IDirect3DSurface d3dInteropSurface;

        // System resources for cube geometry.
        private ModelConstantBuffer modelConstantBufferData;
        private int indexCount = 0;

        // Variables used with the rendering loop.
        private bool loadingComplete = false;

        // If the current D3D Device supports VPRT, we can avoid using a geometry
        // shader just to set the render target array index.
        private bool usingVprtShaders = false;

        private MediaPlayer m_mediaPlayer;
        private Object thisLock = new Object();

        /// <summary>
        /// Loads vertex and pixel shaders from files and instantiates the cube geometry.
        /// </summary>
        public VideoRenderer(DeviceResources deviceResources)
        {
            this.deviceResources = deviceResources;
            m_mediaPlayer = AppView.mediaPlayer;

        }

        /// <summary>
        /// Called once per frame, rotates the cube and calculates the model and view matrices.
        /// </summary>
        public void Update(StepTimer timer)
        {
            Vector3 scale = new Vector3(10.0f);
            Matrix4x4 modelscale = Matrix4x4.CreateScale(scale);

            // Here, we provide the model transform for the sample hologram. The model transform
            // matrix is transposed to prepare it for the shader.
            this.modelConstantBufferData.model = Matrix4x4.Transpose(modelscale);
            // Loading is asynchronous. Resources must be created before they can be updated.
            if (!loadingComplete)
            {
                return;
            }

            // Use the D3D device context to update Direct3D device-based resources.
            var context = this.deviceResources.D3DDeviceContext;

            // Update the model transform buffer for the hologram.
            context.UpdateSubresource(ref this.modelConstantBufferData, this.modelConstantBuffer);
        }

        /// <summary>
        /// Renders one frame using the vertex and pixel shaders.
        /// On devices that do not support the D3D11_FEATURE_D3D11_OPTIONS3::
        /// VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature,
        /// a pass-through geometry shader is also used to set the render 
        /// target array index.
        /// </summary>
        public void Render()
        {
            // Loading is asynchronous. Resources must be created before drawing can occur.
            if (!this.loadingComplete)
            {
                return;
            }

            var context = this.deviceResources.D3DDeviceContext;

            // Each vertex is one instance of the VertexPositionColor struct.
            int stride = SharpDX.Utilities.SizeOf<VertexPositionTexture>();
            int offset = 0;
            var bufferBinding = new SharpDX.Direct3D11.VertexBufferBinding(this.vertexBuffer, stride, offset);
            context.InputAssembler.SetVertexBuffers(0, bufferBinding);
            context.InputAssembler.SetIndexBuffer(
                this.indexBuffer,
                SharpDX.DXGI.Format.R32_UInt, // Each index is one 16-bit unsigned integer (short).
                0);
            context.InputAssembler.PrimitiveTopology = SharpDX.Direct3D.PrimitiveTopology.TriangleList;
            context.InputAssembler.InputLayout = this.inputLayout;

            // Attach the vertex shader.
            context.VertexShader.SetShader(this.vertexShader, null, 0);
            // Apply the model constant buffer to the vertex shader.
            context.VertexShader.SetConstantBuffers(0, this.modelConstantBuffer);

            if (!this.usingVprtShaders)
            {
                // On devices that do not support the D3D11_FEATURE_D3D11_OPTIONS3::
                // VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature,
                // a pass-through geometry shader is used to set the render target 
                // array index.
                context.GeometryShader.SetShader(this.geometryShader, null, 0);
            }

            // Attach the pixel shader.
            context.PixelShader.SetShader(this.pixelShader, null, 0);

            lock (thisLock)
            {

                context.PixelShader.SetShaderResources(0, 1, m_textureView);

                context.PixelShader.SetSamplers(0, 1, m_quadTextureSamplerState);

                // Draw the objects.
                context.DrawIndexedInstanced(
                    indexCount,     // Index count per instance.
                    2,              // Instance count.
                    0,              // Start index location.
                    0,              // Base vertex location.
                    0               // Start instance location.
                    );
            }
        }

        /// <summary>
        /// Creates device-based resources to store a constant buffer, cube
        /// geometry, and vertex and pixel shaders. In some cases this will also 
        /// store a geometry shader.
        /// </summary>
        public void CreateDeviceDependentResourcesAsync()
        {
            ReleaseDeviceDependentResources();

            // Create a constant buffer to store the model matrix.
            modelConstantBuffer = this.ToDispose(SharpDX.Direct3D11.Buffer.Create(
                deviceResources.D3DDevice,
                SharpDX.Direct3D11.BindFlags.ConstantBuffer,
                ref modelConstantBufferData));

            // Create the Texture, ShaderResource and Sampler state
            lock (thisLock)
            {
                m_texture = this.ToDispose(new SharpDX.Direct3D11.Texture2D(deviceResources.D3DDevice, new SharpDX.Direct3D11.Texture2DDescription()
                {
                    Format = SharpDX.DXGI.Format.R8G8B8A8_UNorm,
                    Width = (int)m_mediaPlayer.PlaybackSession.NaturalVideoWidth,//Width
                    Height = (int)m_mediaPlayer.PlaybackSession.NaturalVideoHeight,//Height
                    MipLevels = 1,
                    ArraySize = 1,
                    BindFlags = SharpDX.Direct3D11.BindFlags.ShaderResource | SharpDX.Direct3D11.BindFlags.RenderTarget,
                    Usage = SharpDX.Direct3D11.ResourceUsage.Default,
                    CpuAccessFlags = SharpDX.Direct3D11.CpuAccessFlags.None,
                    SampleDescription = new SharpDX.DXGI.SampleDescription(1, 0),
                    OptionFlags = SharpDX.Direct3D11.ResourceOptionFlags.None
                }));

                m_textureView = this.ToDispose(new SharpDX.Direct3D11.ShaderResourceView(deviceResources.D3DDevice, m_texture));
            }

            m_quadTextureSamplerState = this.ToDispose(new SharpDX.Direct3D11.SamplerState(deviceResources.D3DDevice, new SharpDX.Direct3D11.SamplerStateDescription()
            {
                Filter = SharpDX.Direct3D11.Filter.Anisotropic,
                AddressU = SharpDX.Direct3D11.TextureAddressMode.Clamp,
                AddressV = SharpDX.Direct3D11.TextureAddressMode.Clamp,
                AddressW = SharpDX.Direct3D11.TextureAddressMode.Clamp,
                MaximumAnisotropy = 3,
                MinimumLod = 0,
                MaximumLod = 3,
                MipLodBias = 0,
                BorderColor = new SharpDX.Mathematics.Interop.RawColor4(0, 0, 0, 0),
                ComparisonFunction = SharpDX.Direct3D11.Comparison.Never
            }));

            CreateD3D11Surface();
            LoadShaders();
        }


        private async void LoadShaders()
        {
            usingVprtShaders = deviceResources.D3DDeviceSupportsVprt;
            // On devices that do support the D3D11_FEATURE_D3D11_OPTIONS3::
            // VPAndRTArrayIndexFromAnyShaderFeedingRasterizer optional feature
            // we can avoid using a pass-through geometry shader to set the render
            // target array index, thus avoiding any overhead that would be 
            // incurred by setting the geometry shader stage.
            var vertexShaderFileName = usingVprtShaders ? "Content\\Shaders\\VPRTVertexShader.cso" : "Content\\Shaders\\VertexShader.cso";
            var folder = Windows.ApplicationModel.Package.Current.InstalledLocation;
            // Load the compiled vertex shader.
            var vertexShaderByteCode = await DirectXHelper.ReadDataAsync(await folder.GetFileAsync(vertexShaderFileName));

            // After the vertex shader file is loaded, create the shader and input layout.
            vertexShader = this.ToDispose(new SharpDX.Direct3D11.VertexShader(
                deviceResources.D3DDevice,
                vertexShaderByteCode));

            SharpDX.Direct3D11.InputElement[] vertexDesc =
            {
                new SharpDX.Direct3D11.InputElement("POSITION", 0, SharpDX.DXGI.Format.R32G32B32_Float,  0, 0, SharpDX.Direct3D11.InputClassification.PerVertexData, 0),
                new SharpDX.Direct3D11.InputElement("TEXCOORD", 1, SharpDX.DXGI.Format.R32G32_Float, 12, 0, SharpDX.Direct3D11.InputClassification.PerVertexData, 0),
            };


            inputLayout = this.ToDispose(new SharpDX.Direct3D11.InputLayout(
                deviceResources.D3DDevice,
                vertexShaderByteCode,
                vertexDesc));


            if (!usingVprtShaders)
            {
                // Load the compiled pass-through geometry shader.
                var geometryShaderByteCode = await DirectXHelper.ReadDataAsync(await folder.GetFileAsync("Content\\Shaders\\GeometryShader.cso"));

                // After the pass-through geometry shader file is loaded, create the shader.
                geometryShader = this.ToDispose(new SharpDX.Direct3D11.GeometryShader(
                    deviceResources.D3DDevice,
                    geometryShaderByteCode));
            }

            // Load the compiled pixel shader.
            var pixelShaderByteCode = await DirectXHelper.ReadDataAsync(await folder.GetFileAsync("Content\\Shaders\\PixelShader.cso"));

            // After the pixel shader file is loaded, create the shader.
            pixelShader = this.ToDispose(new SharpDX.Direct3D11.PixelShader(
                deviceResources.D3DDevice,
                pixelShaderByteCode));

            ComputeSphere(16, true);
            // Once the cube is loaded, the object is ready to be rendered.
            loadingComplete = true;
        }

        /// <summary>
        /// Releases device-based resources.
        /// </summary>
        public void ReleaseDeviceDependentResources()
        {
            loadingComplete = false;
            usingVprtShaders = false;
            this.RemoveAndDispose(ref vertexShader);
            this.RemoveAndDispose(ref inputLayout);
            this.RemoveAndDispose(ref pixelShader);
            this.RemoveAndDispose(ref geometryShader);
            this.RemoveAndDispose(ref modelConstantBuffer);
            this.RemoveAndDispose(ref vertexBuffer);
            this.RemoveAndDispose(ref indexBuffer);
            this.RemoveAndDispose(ref m_texture);
            this.RemoveAndDispose(ref m_textureView);
            this.RemoveAndDispose(ref m_quadTextureSamplerState);
        }

        private void CreateD3D11Surface()
        {
            SharpDX.DXGI.Surface2 surface = m_texture.QueryInterface<SharpDX.DXGI.Surface2>();
            IntPtr pUnknown;
            UInt32 hr = InteropStatics.CreateDirect3D11SurfaceFromDXGISurface(surface.NativePointer, out pUnknown);

            if (hr == 0)
            {
                d3dInteropSurface = (IDirect3DSurface)Marshal.GetObjectForIUnknown(pUnknown);
                Marshal.Release(pUnknown);
            }

            m_mediaPlayer.VideoFrameAvailable += M_mediaPlayer_VideoFrameAvailable;
            m_mediaPlayer.Play();
        }

        private void M_mediaPlayer_VideoFrameAvailable(MediaPlayer sender, object args)
        {
            lock (thisLock)
            {
                m_mediaPlayer.CopyFrameToVideoSurface(d3dInteropSurface);
            }
        }

        private void ComputeSphere(int tessellation, bool invertn)
        {
            if (tessellation < 3)
            {
                new Exception("tesselation parameter out of range");
            }

            int verticalSegments = tessellation;
            int horizontalSegments = tessellation * 2;
            int vertCount = (verticalSegments + 1) * (horizontalSegments + 1);
            int idxCount = (verticalSegments) * (horizontalSegments + 1) * 6;

            float radius = 0.5f; // Diameter of the default Sphere will always be 1 to stay aligned


            VertexPositionTexture[] sphereVertArray = new VertexPositionTexture[vertCount];
            int[] cubeIndices = new int[idxCount];

            int writePos = 0;

            for (int i = 0; i <= verticalSegments; i++)
            {
                float v = 1 - (float)i / verticalSegments;

                float latitude = (float)(i * SharpDX.MathUtil.Pi /
                                           verticalSegments) - SharpDX.MathUtil.PiOverTwo;

                var dy = (float)Math.Sin(latitude);
                var dxz = (float)Math.Cos(latitude);

                // Create a single ring of vertices at this latitude.
                for (ushort j = 0; j <= horizontalSegments; j++)
                {
                    float u = (float)j / horizontalSegments;

                    float longitude = (float)(j * SharpDX.MathUtil.TwoPi / horizontalSegments);
                    float dx = (float)Math.Cos(longitude);
                    float dz = (float)Math.Sin(longitude);
                    dx *= dxz;
                    dz *= dxz;

                    Vector3 vect3 = new Vector3(dx, dy, dz);
                    VertexPositionTexture vert;
                    vert.pos = vect3 * radius;
                    vert.texture = new Vector2(u, v);

                    sphereVertArray[writePos] = vert;
                    writePos++;
                }
            }

            // Fill the index buffer with triangles joining each pair of latitude rings.
            int stride = horizontalSegments + 1;
            writePos = 0;
            for (int i = 0; i < verticalSegments; i++)
            {
                for (int j = 0; j <= horizontalSegments; j++)
                {
                    int nextI = i + 1;
                    int nextJ = (j + 1) % stride;

                    cubeIndices[writePos] = i * stride + j;
                    cubeIndices[++writePos] = nextI * stride + j;
                    cubeIndices[++writePos] = i * stride + nextJ;

                    cubeIndices[++writePos] = i * stride + nextJ;
                    cubeIndices[++writePos] = nextI * stride + j;
                    cubeIndices[++writePos] = nextI * stride + nextJ;
                    writePos++;
                }
            }

            vertexBuffer = this.ToDispose(SharpDX.Direct3D11.Buffer.Create(
               deviceResources.D3DDevice,
               SharpDX.Direct3D11.BindFlags.VertexBuffer,
               sphereVertArray));

            indexCount = idxCount;

            indexBuffer = this.ToDispose(SharpDX.Direct3D11.Buffer.Create(
               deviceResources.D3DDevice,
               SharpDX.Direct3D11.BindFlags.IndexBuffer,
               cubeIndices));
        }

        private void XMScalarSinCos(ref float Sin, ref float Cos, float Value)
        {
            const float XM1div2PI = 0.159154943f;
            // Map Value to y in [-pi,pi], x = 2*pi*quotient + remainder.
            float quotient = XM1div2PI * Value;
            if (Value >= 0.0f)
            {
                quotient = (float)((int)(quotient + 0.5f));
            }
            else
            {
                quotient = (float)((int)(quotient - 0.5f));
            }
            float y = Value - SharpDX.MathUtil.TwoPi * quotient;

            // Map y to [-pi/2,pi/2] with sin(y) = sin(Value).
            float sign;
            if (y > SharpDX.MathUtil.PiOverTwo)
            {
                y = SharpDX.MathUtil.Pi - y;
                sign = -1.0f;
            }
            else if (y < -(SharpDX.MathUtil.PiOverTwo))
            {
                y = -(SharpDX.MathUtil.Pi) - y;
                sign = -1.0f;
            }
            else
            {
                sign = +1.0f;
            }

            float y2 = y * y;

            // 11-degree minimax approximation
            Sin = (((((-2.3889859e-08f * y2 + 2.7525562e-06f) * y2 - 0.00019840874f) * y2 + 0.0083333310f) * y2 - 0.16666667f) * y2 + 1.0f) * y;

            // 10-degree minimax approximation
            float p = (float)((float)((((-2.6051615e-07f * y2 + 2.4760495e-05f) * y2 - 0.0013888378f) * y2 + 0.041666638f) * y2 - 0.5f) * y2 + 1.0f);

            Cos = sign * p;
        }

    }
}


